pints/_mcmc/__init__.py

```diff
@@ -12,6 +12,108 @@
 import numpy as np
 
 
+class SingleChainMCMC(object):
+    """
+    Base class for MCMC methods that take a :class:`LogPDF` function and
+    generate a single markov chain representative of its distribution, using an
+    ask-and-tell interface.
+
+    Arguments:
+
+    ``x0``
+        An starting point in the parameter space.
+    ``fx0``
+        The log-pdf of ``x0``.
+    ``sigma0=None``
+        An optional initial covariance matrix, i.e., a guess of the covariance
+        of the distribution to estimate, around ``x0``.
+
+    """
+    def __init__(self, x0, fx0, sigma0=None):
+       
+        # Check initial position
+        self._x0 = pints.vector(x0)
+        
+        # Get dimension
+        self._dimension = len(self._x0)
+        
+        # Check initial log-pdf of x0
+        self._fx0 = float(fx0)
+        if not np.isfinite(fx0):
+            raise ValueError('Initial fx0 must be finite.')
+
+        # Check initial standard deviation
+        if sigma0 is None:
+            self._sigma0 = np.diag(0.01 * self._x0)
+        else:
+            self._sigma0 = np.array(sigma0)
+            if np.product(self._sigma0.shape) == self._dimension:
+                # Convert from 1d array
+                self._sigma0 = self._sigma0.reshape((self._dimension,))
+                self._sigma0 = np.diag(self._sigma0)
+            else:
+                # Check if 2d matrix of correct size
+                self._sigma0 = self._sigma0.reshape(
+                    (self._dimension, self._dimension))
+            self._sigma0.setflags(write=False)
+
+    def ask(self):
+        """
+        Returns a position in the search space to evaluate.
+        """
+        raise NotImplementedError
+        
+    def name(self):
+        """
+        Returns this method's full name.
+        """
+        raise NotImplementedError
+   
+    def tell(self, fx):
+        """
+        Performs an iteration of the MCMC algorithm, using the evaluation
+        ``fx`` of the point ``x`` previously specified by ``ask``. Returns the
+        sample ``x`` accepted by the method.
+        """
+        raise NotImplementedError
+
+
+class SingleChainAdaptiveMCMC(SingleChainMCMC):
+    """
+    *Extends:* :class:`SingleChainMCMC`
+    
+    Base class for adaptive MCMC methods.
+    """
+    def __init__(self, x0, fx0, sigma0=None):
+        super(SingleChainAdaptiveMCMC, self).__init__(x0, fx0, sigma0)
+        self._adaptation = False
+    
+    def adaptation(self):
+        """
+        Returns ``True`` if (and only if) this sampler is currently in
+        adaptation mode.
+        """
+        return self._adaptation
+    
+    def set_adaptation(self, adaptation):
```

Comment:
Yes, something like this! The modal approach seems to involve a lot of code, internal state variable, two separate functions. It might be simpler to have only a `do_adaption_now` type method? Up to you, I'm not sure what is the best way to go